collectors/recon.go

```diff
@@ -0,0 +1,624 @@
+		return
+	}
+
+	gotErr := false
+	for _, match := range matchList {
+		kind := strings.ReplaceAll(string(match[1]), ".", "")
+		matchedHosts, err := strconv.ParseFloat(string(match[2]), 64)
+		if err != nil {
+			logg.Error("recon collector: %s: %v", taskName, err)
+			gotErr = true
+			continue
+		}
+		totalHosts, err := strconv.ParseFloat(string(match[3]), 64)
+		if err != nil {
+			logg.Error("recon collector: %s: %v", taskName, err)
+			gotErr = true
+			continue
+		}
+		notMatchedHosts := totalHosts - matchedHosts
+		errsEncountered, err := strconv.ParseFloat(string(match[4]), 64)
+		if err != nil {
+			logg.Error("recon collector: %s: %v", taskName, err)
+			gotErr = true
+			continue
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			clusterMD5AllDesc,
+			prometheus.GaugeValue, matchedHosts+notMatchedHosts+errsEncountered,
+			kind,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			clusterMD5MatchedDesc,
+			prometheus.GaugeValue, matchedHosts,
+			kind,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			clusterMD5NotMatchedDesc,
+			prometheus.GaugeValue, notMatchedHosts,
+			kind,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			clusterMD5ErrorsDesc,
+			prometheus.GaugeValue, errsEncountered,
+			kind,
+		)
+	}
+	if gotErr {
+		errCount++
+	}
+}
+
+var (
+	clusterCntrUpdaterSweepTimeDesc = prometheus.NewDesc(
+		"swift_cluster_containers_updater_sweep_time",
+		"Container updater sweep time reported by the swift-recon tool.",
+		[]string{"storage_ip"}, nil,
+	)
+	clusterObjUpdaterSweepTimeDesc = prometheus.NewDesc(
+		"swift_cluster_objects_updater_sweep_time",
+		"Object updater sweep time reported by the swift-recon tool.",
+		[]string{"storage_ip"}, nil,
+	)
+)
+
+func reconUpdaterSweepTask(taskName, pathToReconExecutable string, ch chan<- prometheus.Metric) {
+	serverTypes := []string{"container", "object"}
+	for _, server := range serverTypes {
+		errCount := 0
+		cmdArgs := []string{server, "--updater", "--verbose"}
+
+		result, err := getSwiftReconOutputPerHost(pathToReconExecutable, cmdArgs...)
+		if err != nil {
+			logg.Error("recon collector: %s: %s: %v", taskName, server, err)
+			errCount++
+			return
+		}
+
+		gotErr := false
+		for hostname, dataBytes := range result {
+			var data struct {
+				ContainerUpdaterSweepTime float64 `json:"container_updater_sweep"`
+				ObjectUpdaterSweepTime    float64 `json:"object_updater_sweep"`
+			}
+			err := json.Unmarshal(dataBytes, &data)
+			if err != nil {
+				logg.Error("recon collector: %s: %s: %s: %v", taskName, server, hostname, err)
+				gotErr = true
+				continue
+			}
+
+			val := data.ContainerUpdaterSweepTime
+			desc := clusterCntrUpdaterSweepTimeDesc
+			if server == "object" {
+				val = data.ObjectUpdaterSweepTime
+				desc = clusterObjUpdaterSweepTimeDesc
+			}
+
+			ch <- prometheus.MustNewConstMetric(
+				desc,
+				prometheus.GaugeValue, val,
+				hostname,
+			)
+		}
+		if gotErr {
+			errCount++
+		}
+
+		sendReconErrorCount(ch, cmdArgs, errCount)
```

Comment:
This will not be called when the `return` is taken on line 314.

Reply:
There was also another problem with this approach. When `defer sendReconErrorCount()` was used at the top, the value of `errCount` was also evaluated and this meant that a later `errCount++` would have had no effect.

I have fixed this by wrapping these calls in an anonymous function.

While fixing it, I got an idea for a better approach for error handling. I will implement it along with the error test cases in a separate PR. For now, this should suffice.